modules/update-mco-process.adoc

```diff
@@ -0,0 +1,25 @@
+// Module included in the following assemblies:
+//
+// * updating/understanding_updates/how-updates-work.adoc
+
+:_content-type: CONCEPT
+[id="mco-update-process_{context}"]
+= Machine Config Operator node updates
+The Machine Config Operator (MCO) applies a new machine configuration to each control plane and compute node. During this process, the MCO performs the following sequential actions in the cluster:
+
+. Cordon and drain all of the nodes
```

Comment:
should we mention one by one?  
also, a paused machine pool in EUS-to-EUS for example, is an exception.

Reply:
Could you clarify the exact order of operations here? Is it "Cordon and drain each node, one at a time, and _then_ update the OS of each node one at a time, and _then_ reboot one a time, etc."? Or is a single node cordoned, drained, updated, rebooted, uncordoned, and scheduled before the next node initiates the same process?

Reply:
It's not one at a time, it's configurable for each pool (`.spec.maxUnavailable`). One is the default behavior. 

>  Or is a single node cordoned, drained, updated, rebooted, uncordoned, and scheduled before the next node initiates the same process?
^^^ this is the process when maxUnavailable is one

Reply:
I think there's a card for MCO team to provide detailed documentation on their part of the upgrade

Reply:
I think linking to more fleshed out documentation about this specific part would be perfect for a page like this. In the meantime, I can try and rewrite this section to specify that this happens to one pool of cordoned nodes at a time and see if you guys think it looks good/accurate

Reply:
> ...one pool of cordoned nodes at a time...

MachineConfigPools have [a `maxUnavailable` knob, defaulting to 1][1].  When the machine-config controller notices that a MachineConfigPool `spec` calls for an update, and fewer than `maxUnavailable` Nodes in that pool are unavailable, it will cordon that Node, [take some actions which may, but do not necessarily, include draining and/or rebooting][1], and uncordon the Node.  For OCP updates, there are almost always RHCOS updates getting rolled out, and those need the full drain and reboot.  But every once and a while we'll ship the same RHCOS and relevant MachineConfigs in two separate 4.y.z, and then you can have an OCP update that requires no drains or reboots.  Unclear to me how much of this context is worth including.  Personally, I'm fine with anything including:

* Don't even mention the machine-config operator.
* Mention that the machine-config operator exists, that it updates (unpaused) MachineConfigPools, and that that can take some time.
* Explaining that the time MachineConfigPools can take depends (almost always) on the time it takes to drain them, and also on the number of nodes you've ok'ed draining in parallel via `maxUnavailable`.
* Explaining how to figure out [which MachineConfigPool updates will trigger drains, and which will trigger reboots, and which will trigger neither][1].

But it's "up to `maxUnavailable` nodes from a MachineConfigPool" being cordoned at any given time, not the whole pool.

[1]: https://github.com/openshift/machine-config-operator/blob/5843b7a4b27f15bc3313ce8a41ef6de792081fcb/docs/MachineConfigDaemon.md#rebootless-updates

Reply:
Ah, reading [your comment here][1] I now realize that your "one pool of cordoned nodes" means "up to maxUnavailable nodes from a MachineConfigPool".  I'd rather avoid the word "pool" for that, to keep folks from falling into the same misconception I did and assume that when you said "pool of ... nodes" you mean the whole MachineConfigPool.

[1]: https://github.com/openshift/openshift-docs/pull/60091#discussion_r1211916951

Reply:
agree with Trevor, the wording "pool" may mislead in this context. back to line 10 "Cordon and drain all of the nodes" we may rather remove the word "all"

Reply:
If there aren't any particular rewording suggestions, I can use "group" in place of "pool". I can also remove the word "all" to further avoid confusion.

Reply:
Also, I will save further refinement along the lines of Trevor's first comment for a second iteration of this doc.